llvm/lib/Target/AIE/AIE2PreLegalizerCombiner.cpp

```diff
@@ -89,6 +97,83 @@ AIE2PreLegalizerCombinerImpl::AIE2PreLegalizerCombinerImpl(
 {
 }
 
+/// \returns true if it is possible to combine the below sequence of MIRs
+/// into a COPY.
+/// From : %1:_(<64 x s8>) = G_INTRINSIC intrinsic(@llvm.aie2.v64int8/v32int16)
+///        %2:_(<16 x s32>) = G_BITCAST %1:_(<64 x s8>)
+///        %3:_(s32) = G_CONSTANT i32 0
+///        %4:_(<16 x s32>) = G_INTRINSIC
+///        intrinsic(@llvm.aie2.vshift.I512.I512), %X:_(<16 x s32>), %2:_(<16 x
+///        s32>), %3:_(s32), %3:_(s32)
+/// To :   4%:_(<16 x s32>) = COPY %X
+/// Or even:
+/// From : %1:_(<64 x s8>) = G_INTRINSIC intrinsic(@llvm.aie2.v16int32)
+///        %2:_(s32) = G_CONSTANT i32 0
+///        %3:_(<16 x s32>) = G_INTRINSIC
+///        intrinsic(@llvm.aie2.vshift.I512.I512), %X:_(<16 x s32>), %1:_(<16 x
+///        s32>), %2:_(s32), %2:_(s32)
+/// To :   3%:_(<16 x s32>) = COPY %X
+bool AIE2PreLegalizerCombinerImpl::tryToCombineVectorShiftsByZero(
+    MachineInstr &MI) const {
+
+  const Register SecondSrcReg = MI.getOperand(3).getReg();
+  const Register ThirdSrcReg = MI.getOperand(4).getReg();
+  const Register ShiftAmtSrcReg = MI.getOperand(5).getReg();
+
+  auto IsConstantZeroReg = [&](const Register Reg) {
+    auto Cst = getIConstantVRegValWithLookThrough(Reg, MRI);
+    return Cst && Cst->Value.isZero();
+  };
+
+  if (!IsConstantZeroReg(ThirdSrcReg) || !IsConstantZeroReg(ShiftAmtSrcReg))
+    return false;
+
+  const MachineInstr *SecondDefInst = getDefIgnoringCopies(SecondSrcReg, MRI);
+
+  if (SecondDefInst->getOpcode() == TargetOpcode::G_BITCAST)
+    SecondDefInst =
+        getDefIgnoringCopies(SecondDefInst->getOperand(1).getReg(), MRI);
+
+  auto IsIn512BitsUndef = [](unsigned ID) {
+    switch (ID) {
+    case Intrinsic::aie2_v64int8:
+    case Intrinsic::aie2_v32int16:
+    case Intrinsic::aie2_v16int32:
+      return true;
+    default:
+      break;
+    }
+    return false;
+  };
+
+  if (const GIntrinsic *Intrinisc = dyn_cast<GIntrinsic>(SecondDefInst)) {
+    if (!IsIn512BitsUndef(Intrinisc->getIntrinsicID())) {
+      return false;
```

Comment:
Do we need to check this? I think that if we shift by 0, then the second operand does not matter.